source/manuals/programming-languages/nodejs/testing.html.md.erb

```diff
@@ -0,0 +1,80 @@
+---
+title: Testing Node.js Apps at GDS
+---
+
+# <%= current_page.data.title %>
+
+## Libraries
+
+At GDS we use Mocha as our test runner and Chai as our assertion library.
+These form part of a larger group of commonly used testing libraries, a non-exhaustive list of libraries we use can be found below.
+
+| Library | Purpose |
+|:-------|:-------|
+| [Mocha] | Test Runner |
+| [Chai] | Assertion library (NOTE: We use chai's `expect` assertion style and NOT it's `should` style. |
+| [Sinon] | Used to create spys on methods for specific functions |
+| [Nock] | Used for mocking external http and https calls |
+| [Supertest] | Used to test endpoints on Express.js applications |
+| [Pact] | Used for contract testing |
+| [Cheerio] | Used for testing rendered HTML templates |
+
+## We do not use `chai.should()`
+
+We use `chai.expect()` over `chai.should()` for our test assertions, this is because:
+
+- The `.should` style alters the prototype of all objects across Node. Changing prototypes of native types is generally bad practice.
+- Because `.should` affects global scope, it can create interdependent tests unless `chai.should()` is called in every test file. This can be particularly troublesome if you are running/debugging tests in an IDE.
+
+## Testing Asynchronously
+
+Asynchronous code is tested using Mocha's [done] callback. More on this can be found here.
+
+We follow the pattern of handling asynchronous calls in `before()` or `beforeEach()`, then testing the result synchronously and atomicly in subsequent `it()` functions.
+This pattern allows means:
+
+- Its easy to write simple `it()` functions and simple assertions, which are both more readable and more understandable
+- Because `it()` functions can be more atomic in what they test, the feedback when tests brake can be be more descriptive
+
+However it is important to note that any unintended errors thrown in any asynchronous tests are passed to `done()`.
+This means that the content of the errors can be fed back to a developer, so that the developer is not just presented with generic error text such as: `Error: Timeout of 2000ms exceeded`.
+
+An example of this pattern can be found below.
+
+```
+'use strict'
+const request = require('request')
+const {expect} = require('chai')
```

Comment:
Is it worth adding a comment here saying that this is a destructuring pattern? For those not that familiar with ES2015+ it may be helpful to read up on this syntax.

Reply:
Added 👍